File: android/crewrp-core/src/main/kotlin/app/crewrp/core/AuthFlow.kt
package app.crewrp.core

data class AuthConfig(
    val clientId: String,
    val redirectUri: String,
    val authBridgeBaseUrl: String,
)

data class LoginChallenge(
    val authorizeUrl: String,
    val state: String,
    val codeVerifier: String,
)

class AuthFlow(
    private val config: AuthConfig,
    private val bridge: AuthBridgeClient,
    private val membership: GitHubMembershipClient,
    private val tokens: TokenStore,
    private val cache: CacheStore,
) {
    private var pending: LoginChallenge? = null

    fun beginLogin(): LoginChallenge {
        val pkce = PKCE.generate()
        val state = PKCE.generate().verifier
        val url = GitHubOAuth.authorizeUrl(
            clientId = config.clientId,
            redirectUri = config.redirectUri,
            state = state,
            codeChallenge = pkce.challenge,
        )
        return LoginChallenge(url, state, pkce.verifier).also { pending = it }
    }

    fun completeLogin(callbackUrl: String): List<CrewRepo> {
        val query = URIQuery.parse(callbackUrl)
        val pending = pending ?: error("missing pending login")
        require(query["state"] == pending.state) { "state mismatch" }
        val code = query["code"] ?: error("missing code")
        val token = bridge.exchange(code, pending.codeVerifier, config.redirectUri)
        tokens.saveAccessToken(token.accessToken)
        this.pending = null
        val repos = membership.listRegistrableRepos(token.accessToken)
        require(repos.isNotEmpty()) { "no registrable repos" }
        return repos
    }

    fun registerCrew(repo: CrewRepo): Session {
        val token = tokens.loadAccessToken() ?: error("missing token")
        val role = membership.resolveRole(repo.owner, token, isRepoAdmin = true)
        val session = Session(repo.owner, repo.fullName, role)
        cache.putSession(session)
        return session
    }
}

internal object URIQuery {
    fun parse(url: String): Map<String, String> {
        val q = url.substringAfter('?', missingDelimiterValue = "")
        if (q.isEmpty()) return emptyMap()
        return q.split('&').mapNotNull { part ->
            val idx = part.indexOf('=')
            if (idx < 0) null
            else {
                val key = java.net.URLDecoder.decode(part.substring(0, idx), Charsets.UTF_8)
                val value = java.net.URLDecoder.decode(part.substring(idx + 1), Charsets.UTF_8)
                key to value
            }
        }.toMap()
    }
}

File: ios/Sources/CrewRPCore/AuthFlow.swift
import Foundation

public struct AuthConfig: Sendable, Equatable {
    public var clientID: String
    public var redirectURI: String
    public var authBridgeBaseURL: URL

    public init(
        clientID: String,
        redirectURI: String,
        authBridgeBaseURL: URL
    ) {
        self.clientID = clientID
        self.redirectURI = redirectURI
        self.authBridgeBaseURL = authBridgeBaseURL
    }
}

public struct LoginChallenge: Sendable, Equatable {
    public let authorizeURL: URL
    public let state: String
    public let codeVerifier: String
}

public enum AuthFlowError: Error, Equatable {
    case stateMismatch
    case missingCode
    case noRegistrableRepos
}

public final class AuthFlow: @unchecked Sendable {
    private let config: AuthConfig
    private let bridge: AuthBridgeClient
    private let membership: GitHubMembershipClient
    private let tokens: any TokenStore
    private let cache: CacheStore
    private var pending: LoginChallenge?

    public init(
        config: AuthConfig,
        bridge: AuthBridgeClient,
        membership: GitHubMembershipClient,
        tokens: any TokenStore,
        cache: CacheStore
    ) {
        self.config = config
        self.bridge = bridge
        self.membership = membership
        self.tokens = tokens
        self.cache = cache
    }

    public func beginLogin() -> LoginChallenge {
        let pkce = PKCE.generate()
        let state = PKCE.generate().verifier
        let url = GitHubOAuth.authorizeURL(
            clientID: config.clientID,
            redirectURI: config.redirectURI,
            state: state,
            codeChallenge: pkce.challenge
        )
        let challenge = LoginChallenge(authorizeURL: url, state: state, codeVerifier: pkce.verifier)
        pending = challenge
        return challenge
    }

    /// Exchanges the OAuth code and returns private repos the user can register as a crew.
    public func completeLogin(callbackURL: URL) async throws -> [CrewRepo] {
        let items = URLComponents(url: callbackURL, resolvingAgainstBaseURL: false)?.queryItems ?? []
        var map: [String: String] = [:]
        for item in items {
            if let value = item.value {
                map[item.name] = value
            }
        }
        guard let pending else { throw AuthFlowError.missingCode }
        guard map["state"] == pending.state else { throw AuthFlowError.stateMismatch }
        guard let code = map["code"] else { throw AuthFlowError.missingCode }

        let token = try await bridge.exchange(
            code: code,
            codeVerifier: pending.codeVerifier,
            redirectURI: config.redirectURI
        )
        try tokens.saveAccessToken(token.accessToken)
        self.pending = nil

        let repos = try await membership.listRegistrableRepos(token: token.accessToken)
        guard !repos.isEmpty else { throw AuthFlowError.noRegistrableRepos }
        return repos
    }

    public func registerCrew(_ repo: CrewRepo) async throws -> Session {
        guard let token = try tokens.loadAccessToken() else { throw AuthFlowError.missingCode }
        let role = try await membership.resolveRole(owner: repo.owner, token: token, isRepoAdmin: true)
        let session = Session(org: repo.owner, repo: repo.fullName, teamRole: role)
        try cache.putSession(session)
        return session
    }
}
